feat: proposers can cancel their proposals (#2272)
vegaprotocol/specs: protocol/0028-GOVE-governance.md

at e.g. `09:00:00 on 1st Jan 2021` and `governance.proposal.asset.minEnact` is `72h` then the proposal must contain enactment date/time that after `09:00:00 on 4th Jan 2021`.
If there is `governance.proposal.asset.maxEnact` of e.g. `360h` then the proposed enactment date / time must be before `09:00:00 on 16th Jan 2021`.

> ## Editing a proposal
> 
> A proposal cannot be edited, once created. The only possible action is to vote for or against a proposal, or submit a new proposal.

Reviewer: ```suggestion
## Editing a proposal

A proposal cannot be edited once submitted to the network. The only possible actions are to cancel within the threshold, vote for or against a proposal, or submit a new proposal.
```

Which of these makes most sense will depend on the type of change, the timing of the events, and how the rest of the community votes for the initial proposal.

## Cancelling a proposal

A proposal can be cancelled under certain conditions by, and only by, the party which created the proposal. To avoid situations where the community may have acted upon proposals they "expected" to pass, a party can only cancel their proposal before a certain "cancellation threshold" is reached.

> This cancellation threshold is a fixed point before the proposal closing time. The duration of the period in which a proposal cannot be cancelled is defined by the network parameter `governance.proposal.cancellationThreshold` (a duration string defaulting to `24h0m0s`).

Reviewer 2: Given that there are proposal types with 24h total time possible, I would suggest making the default for this at least a bit lower—something like 20h or 23h—so that a proposal with even only 1 day of voting at least has some window in which mistakes can be corrected.

Reviewer 3: `23h`.

> This cancellation threshold is a fixed point before the proposal closing time. The duration of the period in which a proposal cannot be cancelled is defined by the network parameter `governance.proposal.cancellationThreshold` (a duration string defaulting to `24h0m0s`).

Reviewer 4: The original issue wanted to make the cancelation period relative to th3e voting period, while the spec now makes it a constant period. This shouldn't be an issue, but might be a bit odd if we have votes with a very short voting time (e.g., validator-only votes for changing spam parameters). In that case the cancelation period could be longer than the voting period, which leads to edge cases (maybe add to the spec something like "if the voting period is shorter than the cancelation period, then the cancelation period is set to the duration of the voting period)

Reviewer 2: I think this is the other way round, so you'd end up unable to cancel a short enough vote, as it's a threshold that says how long before the end you bvecome unable to cancel a proposal, not how long it can be cancelled for.

> i.e. Given the `governance.proposal.cancellationThreshold` is `24h0m0s` a party will be able to cancel their proposal 1 day before the proposal closing time.

Reviewer 5: ```suggestion
i.e. Given the `governance.proposal.cancellationThreshold` is `24h0m0s` a party will be able to cancel their proposal up until 1 day before the proposal closing time.
```

### Mechanics

> A party can cancel their proposal by simply voting against it with any vote value.

Reviewer 2: What does "with any vote value" mean here? I think votes are VALUE_YES and VALUE_NO so you would need to vote VALUE_NO?

- If the current time is before the cancellation threshold, the proposal is instantly rejected.
- If the current time is after the cancellation threshold, the vote is treated normally as a vote against a proposal.

Note the following additional details:

- a cancelling vote against a proposal must pass all typical vote checks and spam protections, i.e. who party who removes stake after creating a proposal and no longer meets the requirement will not be able to cancel their proposal.

> - if at the time of creating a proposal the closing time is within the period defined by `governance.proposalCancellationThreshold`. The proposal will still be valid and accepted by the network but the party will be not have a opportunity to cancel the proposal.

Reviewer: ```suggestion
- if at the time of creating a proposal the closing time is within the period defined by `governance.proposalCancellationThreshold`. The proposal will still be valid and accepted by the network but the party will not have an opportunity to cancel the proposal.
```

> - if at the time of creating a proposal the closing time is within the period defined by `governance.proposal.cancellationThreshold`. The proposal will still be valid and accepted by the network but the party will not have an opportunity to cancel the proposal.

Reviewer: ```suggestion
- if at the time of creating a proposal the closing time is within the period defined by `governance.proposal.cancellationThreshold`, the proposal will still be valid and accepted by the network but the party will not have an opportunity to cancel the proposal.
```

> - if at the time of creating a proposal the closing time is within the period defined by `governance.proposal.cancellationThreshold`, the proposal will still be valid and accepted by the network but the party will not have an opportunity to cancel the proposal.

Reviewer 2: We should mention that their vote will still be counted as (changed to) a vote against even if it is too late to canel the proposal

Author: Yeah I agree, hopefully `line 147` makes this point already though?

- A voter's equity-like share does not give them any additional voting weight when voting on a market community tags update proposal. (<a name="0028-GOVE-174" href="#0028-GOVE-174">0028-GOVE-174</a>)
- A proposal to add community tags with any community tags longer than `governance.proposal.market.maxCommunityTagLength` is rejected as invalid (<a name="0028-GOVE-175" href="#0028-GOVE-175">0028-GOVE-175</a>)
- A proposal to remove community tags with any community tags longer than `governance.proposal.market.maxCommunityTagLength` is rejected as invalid (<a name="0028-GOVE-176" href="#0028-GOVE-176">0028-GOVE-176</a>)

#### Cancelling Proposals

> - Given a proposal with a closing time further than `governance.proposal.cancellationThreshold` from the current time, if the party which created the proposal votes against it, the proposal will be instantaneously rejected (<a name="0028-GOVE-188" href="#0028-GOVE-188">0028-GOVE-188</a>)

Reviewer: ```suggestion
- Given a proposal with a closing time further than `governance.proposal.cancellationThreshold` from the current time, if the party that created the proposal votes against it, the proposal will be instantaneously rejected (<a name="0028-GOVE-188" href="#0028-GOVE-188">0028-GOVE-188</a>)
```

> - Given a proposal with a closing time less than `governance.proposal.cancellationThreshold` from the current time, if the party which created the proposal votes against it, the proposal will not be instantaneously rejected. At closing time, if the proposal meets the criteria to be accepted it will pass (<a name="0028-GOVE-189" href="#0028-GOVE-189">0028-GOVE-189</a>)

Reviewer: ```suggestion
- Given a proposal with a closing time less than `governance.proposal.cancellationThreshold` from the current time, if the party that created the proposal votes against it, the proposal will not be instantaneously rejected. At closing time, if the proposal meets the criteria to be accepted it will pass (<a name="0028-GOVE-189" href="#0028-GOVE-189">0028-GOVE-189</a>)
```

> - Given a proposal with a closing time less than `governance.proposal.cancellationThreshold` from the current time, if the party which created the proposal votes against it, the proposal will not be instantaneously rejected. At closing time, if the proposal does not meet the criteria to be accepted it will be rejected (<a name="0028-GOVE-190" href="#0028-GOVE-190">0028-GOVE-190</a>)

Reviewer: ```suggestion
- Given a proposal with a closing time less than `governance.proposal.cancellationThreshold` from the current time, if the party that created the proposal votes against it, the proposal will not be instantaneously rejected. At closing time, if the proposal does not meet the criteria to be accepted it will be rejected (<a name="0028-GOVE-190" href="#0028-GOVE-190">0028-GOVE-190</a>)
```

> - Given a proposal with a closing time further than `governance.proposal.cancellationThreshold` from the current time, if the party which did not create the proposal votes against it, the proposal will not be instantaneously rejected (<a name="0028-GOVE-191" href="#0028-GOVE-191">0028-GOVE-191</a>)

Reviewer: ```suggestion
- Given a proposal with a closing time further than `governance.proposal.cancellationThreshold` from the current time, if a party that did not create the proposal votes against it, the proposal will not be instantaneously rejected (<a name="0028-GOVE-191" href="#0028-GOVE-191">0028-GOVE-191</a>)
```
